File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void func(TreeNode root, List<List<Integer>> ans, List<Integer> path, int sum, int target) {
        if (root == null)
            return;
        sum += root.val;
        path.add(root.val);
        if (root.left == null && root.right == null && sum == target)
            ans.add(new ArrayList<>(path));

        func(root.left, ans, path, sum, target);
        func(root.right, ans, path, sum, target);
        path.remove(path.size() - 1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null)
            return ans;
        func(root, ans, new ArrayList<Integer>(), 0, targetSum);
        return ans;
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> PathSum(TreeNode root, int targetSum) {
        var ans = new List<IList<int>>();
        var path = new List<int>();

        void func(TreeNode root, int sum, int target) {
            if (root == null)
                return ;
            
            path.Add(root.val);
            sum += root.val;

            if (root.left == null && root.right == null && sum == target)
                ans.Add(new List<int>(path));
            func(root.left, sum, target);
            func(root.right, sum, target);
            path.RemoveAt(path.Count - 1);
        }
        
        func(root, 0, targetSum);
        return ans;
    }
}